imx6ull-flash-test: add ECC tests

Maxez opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes shortly -->
Added flash ECC tests (meta and data chunks are checked for correcting maximum recoverable number of bit flips).
Added read disturb stress test.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
[JIRA: DTR-96](https://jira.phoenix-rtos.com/browse/DTR-96)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

<!--- In case of breaking change - please advice here what needs to be done in dependent projects. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: armv7a7-imx6ull.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

Maxez commented:
Expected test output:
![test_ecc](https://user-images.githubusercontent.com/8835317/130631812-72da4564-5463-413a-a8a8-c768303aad74.png)

kemonats commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -305,6 +316,122 @@ void test_bootrom(void)
+		return;
+	}
+
+	if ((sdata = mmap(NULL, (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1), PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0)) == MAP_FAILED) {
```
Maybe instead of using `(FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1)` ( occurs seven times ) define for example such a variable:
```
const unsigned int maplen = (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1);
```
and use it in `mmap()` and `munmap()`.

Maxez replied:
@nalajcie do you agree with this? Maybe we could define PAGE_ALIGN(addr) macro and use it instead? (sidenote: I was instructed not to write such macros before).

Maxez replied:
Added `static size_t pagemapsz` variable storing mapped flash page size. It's initialized in `main()` right after `flashinfo`.

kemonats commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -305,6 +316,122 @@ void test_bootrom(void)
+		/* Flip data bits (ECC should correct up to 14 bit flips) */
+		for (i = 1; i < sizeof(saux->errors); i++) {
+			printf("[%4u]: adding bit flips to data chunk %u (%u bad bits already present)\n", paddr, i, saux->errors[i]);
+			if ((err = flip_bits(maxdataecc - saux->errors[i], nblock, start + rawmetasz + (i - 1) * rawdatasz, start + rawmetasz + i * rawdatasz, 0)) < 0)
+				break;
```
```suggestion
			start += rawmetasz;
			if ((err = flip_bits(maxdataecc - saux->errors[i], nblock, start, start + rawmetasz, 0)) < 0)
				break;
```
It will be simpler.

Maxez replied:
For flipping bits in the data chunks we want to start at offset `start + rawmetasz` and than advance to the next chunk by adding `rawdatasz` (so we'd need `start += rawdatasz`).
However, I don't want to modify `start` variable because it's used later on for restoring flipped bits.

kemonats replied:
My mistake. I mixed up the variables. I meant to suggest using addition instead multiplication.
My suggestion:
1. The `start, nblock, seed` variables can be defined as `const` ( I guess they are not changed ).
2. You cane define a variable:
```
unsigned int starttmp;
```
and use it in `for` loop:
```
starttmp = start + rawmetasz;
for (i = 1; i < sizeof(saux->errors); i++) {
	printf("[%4u]: adding bit flips to data chunk %u (%u bad bits already present)\n", paddr, i, saux->errors[i]);
	if ((err = flip_bits(maxdataecc - saux->errors[i], nblock, starttmp, starttmp + rawdatasz, 0)) < 0)
		break;
		}
	starttmp += rawdatasz;
}
```

Maxez replied:
I've added the proposed changes. I think the code becomes more readable. Thanks.

Maxez commented:
I've noticed that the test doesn't pass for same flash pages - possible bug in flipping metadata bits:
![ecc_meta_fail](https://user-images.githubusercontent.com/8835317/130781066-efc2349b-3e95-407c-a59b-551603bd9ee9.png)
I'm working on a fix.

Maxez commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -305,6 +316,144 @@ void test_bootrom(void)
+	const unsigned int rawmetasz = 16 + 26;             /* 16B META + 26B ECC16 */
+	const unsigned int rawdatasz = 512 + 23;            /* 512B DATA + 22.75B ECC14 (adding 2 bits for byte alignment) */
+	const unsigned int maxmetaecc = 16;                 /* Max number of bits per meta chunk ECC can correct */
+	const unsigned int maxdataecc = 14;                 /* Max number of bits per data chunk ECC can correct */
```
Fixed.

nalajcie commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -499,6 +648,69 @@ void test_badblocks(void)
+/* Continuously reads blocks and checks for appearing bad bits (reads few different blocks to avoid cache usage) */
+void test_read_disturb(void)
+{
+	const unsigned int blocks[] = { 1111, 3333, 5555, 7777 };        /* Tested blocks */
```
should we first program these pages with some meaningful data with ECC?

Maxez replied:
I've run the test a few times already and after many reads erased pages start to report `flash_uncorrectable` error too but I'm not sure how to interpret this result since as you noticed they don't have a valid ECC.
I guess the test should be run on correctly flashed pages so we can be sure of the meaning of the results.
I'll change `test_ecc()` to program tested page too.

nalajcie reviewed (approved):
Approving as it's better to have more and more tests :).

Left some small comments.

Maxez commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -305,6 +316,144 @@ void test_bootrom(void)
+/* Checks ECC corrections reported by the BCH */
+void test_ecc(void)
+{
+	const unsigned int paddr = 8 * BLOCK_PAGES_CNT; /* Page under testing (should be correctly flashed page except for FCB - different configuration) */
```
It should be first FW1 page (starting at block 8) flashed with out BCH settings.

Maxez commented:
@nalajcie I've changed the tests so that they always erase and program the tested blocks. Also added `pagemapsz` variable for flash page mmap size. Please check and aprove the PR again.

nalajcie commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -497,8 +498,8 @@ void test_meta(void)
-	data = mmap(NULL, _PAGE_SIZE * 2, PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0);
-	meta = data + _PAGE_SIZE;
+	data = mmap(NULL, pagemapsz, PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0);
+	meta = data + flashinfo.writesz;
```
not sure if we shouldn't have `_PAGE_SIZE` here as DMA buffers need to be (RAM) page-aligned regardless of `flashinfo.writesz`. There was a silent assumption here that `flashinfo.writesz <= _PAGE_SIZE` (we can probably add assert / check it when starting tests just to be sure).

This affects other places where `mmap`'ing data + meta together.

Maxez replied:
I agree, `meta` should be explicitly initialized to page aligned address. I'm going to add the proposed assert in `main()` and revert this change.

nalajcie commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -727,13 +754,15 @@ void test_stress_one_block(void)
-	data = mmap(NULL, _PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0);
-	m = meta = mmap(NULL, _PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0);
+	data = mmap(NULL, pagemapsz, PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0);
+	meta = data + flashinfo.writesz;
```
`_PAGE_SIZE`?

Maxez replied:
Fixed.

nalajcie commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -937,11 +966,11 @@ void test_write_read_erase(void)
+	if (data == MAP_FAILED)
 		FAIL("failed to mmap data buffers\n");
 
+	meta = data + flashinfo.writesz;
```
`_PAGE_SIZE`?

nalajcie reviewed (commented):
I think we can merge this, I've pointed out some minute details (mostly semantics).

Please squash / rebase fixup commits to the original ones (I think we should keep either 1 or 2 commits in this PR to rebase merge it to master).

Maxez commented on storage/imx6ull-flash/imx6ull-flash-test.c:
```diff
@@ -510,15 +511,15 @@ void test_meta(void)
 		printf("write() failed: %d\n", err);
 
-	memset(data, 0x0, _PAGE_SIZE * 2);
+	memset(data, 0x0, FLASHDRV_PAGESZ);
```
Changed setting both data and meta pages with `memset(data, 0x0, pagemapsz)`.

Maxez commented:
@nalajcie I squashed all the changes into 2 commits. I think the PR is ready for merge now.

nalajcie reviewed (approved):
Yep, LGTM.
